boot/boot.go

```diff
@@ -403,12 +405,79 @@ func makeBootable20(model *asserts.Model, rootdir string, bootWith *BootableSet,
 	return nil
 }
 
+// XXX: kept this here for now until we have a decision of the file names and locations,
+// then we refactor to integrate this functionality to existing infrastructure. (Currently
+// bootloader.ExtractKernelAssets tests for meta/force-kernel-extraction and
+// bootloader.extractKernelAssetsToBootDir extracts to a subdir)
+func extractKernelAssets(dstDir string, s snap.PlaceInfo, snapf snap.Container) error {
+	dir, err := os.Open(dstDir)
+	if err != nil {
+		return err
+	}
+	defer dir.Close()
+
+	for _, src := range []string{"kernel.img", "initrd.img"} {
+		if err := snapf.Unpack(src, dstDir); err != nil {
+			return err
+		}
+		if err := dir.Sync(); err != nil {
+			return err
+		}
+	}
+	if err := snapf.Unpack("dtbs/*", dstDir); err != nil {
+		return err
+	}
+
+	return dir.Sync()
+}
+
+func makeBootableRunMode(model *asserts.Model, rootdir string, bootWith *BootableSet, opts *bootloader.Options) error {
+	// XXX: Create correct grub.cfg in ubuntu-boot
+
+	// Extract the boot kernel to ubuntu-boot
+	kernelf, err := snap.Open(bootWith.KernelPath)
+	if err != nil {
+		return fmt.Errorf("cannot open kernel snap file: %v", err)
+	}
+	if err := extractKernelAssets(rootdir, bootWith.Kernel, kernelf); err != nil {
+		return fmt.Errorf("cannot extract kernel assets: %v", err)
+	}
+
+	// Write modeenv in ubuntu-data
+	logger.Noticef("write modeenv")
+	modeenv := &Modeenv{
+		Mode:           "run",
+		RecoverySystem: filepath.Base(bootWith.RecoverySystemDir),
+		Base:           filepath.Base(bootWith.BasePath),
+		Kernel:         filepath.Base(bootWith.KernelPath),
+	}
+	if err := modeenv.Write(filepath.Join(dirs.MountPointDir, "ubuntu-data", "system-data")); err != nil {
+		return fmt.Errorf("cannot write modeenv: %v", err)
+	}
+
+	// Update the recovery mode for the next boot
```

Comment:
This portion is nice. Can we have something similar done in the `snap prepare-image`, i.e. it should point at the "install" mode, and thus set `snapd_recovery_mode=install snapd_recovery_system=YYYYMMDD` please? (as a separate PR I guess). Because at the moment, I believe, the images are prepared without any toplevel grubenv dictating what the "firstboot" menu item should be.